piccolo/query/mixins.py

```diff
@@ -237,6 +249,56 @@ def copy(self) -> OutputDelegate:
         return self.__class__(_output=_output)
 
 
+@dataclass
+class CallbackDelegate:
+    """
+    Example usage:
+
+    .callback(my_handler_function)
+    .callback(print, on=Callback.success)
+    .callback(my_handler_coroutine)
+    .callback([handler1, handler2])
+    """
+
+    _callbacks: t.List[RegisteredCallback] = field(default_factory=list)
```

Comment:
One suggestion here - I think it works well, but rather than having to loop through to find the right callbacks, it could be something like this instead, so it's easier to directly retrieve the callbacks we're interested in:

```python
@dataclass
class RegisteredCallbacks:
    success: t.List[Callback] = field(default_factory=list)

class CallbackDelegate:
    _callbacks: RegisteredCallbacks = field(default_factory=RegisteredCallbacks)

# When registering
self._callbacks.success.append(some_function)

# So then we can quickly get the callbacks we want
self._callbacks.success
```

Or maybe a dictionary is better:

```python
class CallbackDelegate:
    _callbacks: t.Dict[Callback, t.List[RegisteredCallback]] = field(default_factory={i: [] for i in Callback})

# When registering
self._callbacks[Callback.success].append(some_function)

# So then we can quickly get the callbacks we want
self._callbacks[Callback.success]
```

What do you think about renaming the `Callback` enum to something like `CallbackType`, or `InvokedOn`? Someone might assume `Callback` corresponds to a function.

Reply:
i was thinking the loop & filter was a bit unwieldy. i think i like the dict storage best, otherwise we'll have to map the `kind` argument to the correct `RegisteredCallbacks` property ourselves.

RE changing the name, absolutely agree it's a bit confusing. changing that would also let me change `RegisteredCallback` back to just `Callback`, which read a bit nicer.

Reply:
You're right - I think the dict is cleaner.